util/include/thread_pool.hpp

```diff
@@ -73,11 +74,24 @@ class ThreadPool {
     {
       auto lock = std::lock_guard{task_queue_.mutex};
       task_queue_.tasks.push(std::move(task));
+      auto done_lock = std::lock_guard{task_queue_.done_check_mutex};
+      task_queue_.tcount++;
     }
     task_queue_.cv.notify_one();
     return future;
   }
 
+  // Call this only after putting all jobs or tasks
+  // This is a special function, which wait till all the
+  // tasks are done
+  void finalWaitForAll() {
```

Comment:
Nit:
As far as I understand, this approach is designed for a single producer case:
Thread A may lock on `finalWaitForAll` (`task_queue_.done_check_mutex`) while thread B would continue adding tasks via `auto async(F&& func, Args&&... args)`.

Is it expected behavior?

@dartdart26 , would mind having a look?

Reply:
Here this ability will provide a Thread bag facility. So consider that we create a thread pool of 4 threads. Now consider that we have 50 tasks [lambda] to run. So thread A will put all the tasks in the bag / pool of threads in any order. After that thread A will wait until the 4 threads of the thread pool / bag finishes those 50 tasks. So if we have a finite tasks in hand that we want to process, then we will use this to wait for all threads to complete, if we don't have this kind of scenario then we don't have to wait.

Reply:
it seems logically correct but if a bug is hidden then we can encounter a deadlock, since we must have a symmetry between the task increment and decrement.
I also suggest @dartdart26 to review this code

Reply:
@arc-vmware , I am afraid that if there are multiple producer threads, the method `finalWaitForAll` will exit before all the tasks are completed. It seems to me that the method covers a very specific use case and may leave the user with a false impression that it will work for all the cases.

The method `async` returns future so if a user wants to wait until all their tasks are done, they need to save the futures and wait for them, i.e.:

```
std::vector<future> results;
for (...) {
   results.push_back(pool.async( my_func );
 }
 
 for(auto& r: results) {
    f.wait();
}
```

WDYT?